oqsprov/oqsprov_keys.c

```diff
@@ -159,24 +243,61 @@ int oqsx_key_fromdata(OQSX_KEY *key, const OSSL_PARAM params[], int include_priv
 
 int oqsx_key_gen(OQSX_KEY *key)
 {
-    if (key->privkey == NULL || key->pubkey == NULL) oqsx_key_allocate_keymaterial(key);
+    int ret = 0, ret2 = 0;
+
+    ret = oqsx_key_allocate_keymaterial(key);
+    ON_ERR_GOTO(ret, err);
 
-    if (key->iskem)
-        return OQS_KEM_keypair(key->key.k, key->pubkey, key->privkey);
-    else
-        return OQS_SIG_keypair(key->key.s, key->pubkey, key->privkey);
+    if (key->keytype == KEY_TYPE_KEM) {
+        ret = OQS_KEM_keypair(key->primitive.kem, key->pubkey, key->privkey);
+        ON_ERR_GOTO(ret, err);
+    } else if (key->keytype == KEY_TYPE_ECP_HYB_KEM || key->keytype == KEY_TYPE_ECX_HYB_KEM) {
+        OQS_HYB_KEM *hybkem = key->primitive.hybkem;
+        OQS_KEM *kem = hybkem->kem;
+        EVP_PKEY_CTX *kex = hybkem->kex;
+        EVP_PKEY *pkey = NULL;
+
+        size_t privkeykemlen = kem->length_secret_key, privkeykexlen = 0;
+        size_t pubkeykemlen = kem->length_public_key, pubkeykexlen = 0;
+
+        ret = OQS_KEM_keypair(hybkem->kem, key->pubkey + 4, key->privkey + 4);
+        ON_ERR_GOTO(ret, err);
+
+        ret2 = EVP_PKEY_keygen_init(kex);
+        ON_ERR_SET_GOTO(ret2 != 1, ret, 10, err);
+        ret2 = EVP_PKEY_keygen(kex, &pkey);
+        ON_ERR_SET_GOTO(ret2 != 1, ret, 11, err);
+
+        ret2 = EVP_PKEY_get_raw_public_key(pkey, key->pubkey + 4 + kem->length_public_key + 4, &pubkeykexlen);
+        ON_ERR_SET_GOTO(ret2 <= 0, ret, 44, err);
+
+        ret2 = EVP_PKEY_get_raw_private_key(pkey, key->privkey + 4 + kem->length_secret_key + 4, &privkeykexlen);
+        ON_ERR_SET_GOTO(ret2 <= 0, ret, 45, err);
+
+        ENCODE_UINT32((unsigned char *)key->pubkey, pubkeykemlen);
+        ENCODE_UINT32((unsigned char *)key->pubkey + 4 + pubkeykemlen, pubkeykexlen);
+        ENCODE_UINT32((unsigned char *)key->privkey, privkeykemlen);
+        ENCODE_UINT32((unsigned char *)key->privkey + 4 + privkeykemlen, privkeykexlen);
+    } else if (key->keytype == KEY_TYPE_SIG) {
+        ret = OQS_SIG_keypair(key->primitive.sig, key->pubkey, key->privkey);
+        ON_ERR_GOTO(ret, err);
+    } else {
+        ret = 1;
+    }
+    err:
+    return ret;
 }
 
 int oqsx_key_parambits(OQSX_KEY *key) {
-    if (key->iskem) 
-        return 128+(key->key.k->claimed_nist_level-1)/2*64;
-    return 128+(key->key.s->claimed_nist_level-1)/2*64;
+    if (key->keytype == KEY_TYPE_KEM)
+        return 128+(key->primitive.kem->claimed_nist_level-1)/2*64;
+    else if (key->keytype == KEY_TYPE_ECP_HYB_KEM || key->keytype == KEY_TYPE_ECX_HYB_KEM)
+        return 128+(key->primitive.hybkem->kem->claimed_nist_level-1)/2*64;
+    return 128+(key->primitive.sig->claimed_nist_level-1)/2*64;
 }
 
 int oqsx_key_maxsize(OQSX_KEY *key) {
-    if (key->iskem)
-        return key->key.k->length_shared_secret;
-    return key->key.s->length_signature;
+    if (key->keytype == KEY_TYPE_KEM)
+        return key->primitive.kem->length_shared_secret;
```

Comment:
Shouldn't here also be an `elseif` on the new hybs?

Reply:
Yes correct, the `elseif` for the hybrids is still missing.

Reply:
added in 0576ca5b5d5c9ca855aa41bb0cc0310e1e48af95